src/safeposix/syscalls/fs_calls.rs

```
@@ -3183,57 +3276,78 @@ impl Cage {
                             }
                             let filesize = normalfile_inode_obj.size;
+                            //The offset cannot be negative, and we cannot read past the end of the file
                             if off < 0 || off > filesize as i64 {
                                 return syscall_error(Errno::ENXIO, "mmap", "Addresses in the range [off,off+len) are invalid for the object specified by fildes.");
                             }
-                            //because of NaCl's internal workings we must allow mappings to extend past the end of a file
+                            //Because of NaCl's internal workings we must allow mappings to extend past the end of the file
                             let fobj = FILEOBJECTTABLE.get(&normalfile_filedesc_obj.inode).unwrap();
-                            //we cannot mmap a rust file in quite the right way so we retrieve the fd number from it
-                            //this is the system fd number--the number of the lind.<inodenum> file in our host system
+                            //The actual memory mapping is not emulated inside Lind, so the call to the kernell
+                            //is required. To perform this call, the file descriptor of the actual file
+                            //stored on the host machine is needed. Since Lind's emulated filesystem
+                            //does not match the underlying host's filesystem, the file descriptor
+                            //provided to the `mmap_syscall()` cannot be used and must be converted
+                            //to the actual file descriptor stored in the host's filesystem.
                             let fobjfdno = fobj.as_fd_handle_raw_int();
-
-
                             interface::libc_mmap(addr, len, prot, flags, fobjfdno, off)
                         }
-
                         Inode::CharDev(_chardev_inode_obj) => {
```

We can move this up the order to fail early.